File: server/lib/comparer.js
'use strict';
// 多篇论文对比：技术路线演化、优缺点、Research Gap。
const { getProvider } = require('./llm');
const store = require('./store');
const { extractJson, truncate } = require('./util');

function paperSummary(paperId) {
  const meta = store.readJson(paperId, 'metadata.json', {});
  const analysis = store.readJson(paperId, 'analysis.json', {});
  const sec = (analysis && analysis.sections) || {};
  return {
    paper_id: paperId,
    title: meta.title || analysis.title || paperId,
    year: meta.year || analysis.year || null,
    research_problem: sec.research_problem || '',
    proposed_method: sec.proposed_method || '',
    results: sec.results || '',
    contributions: sec.contributions || [],
    limitations: sec.limitations || [],
    research_gap: sec.research_gap || [],
  };
}

function buildComparisonPrompt(summaries) {
  const lines = [
    'Compare the following papers and return ONLY valid JSON (no fences) matching this schema:',
    '{ "method_evolution": "...", "strengths_weaknesses": {"<paper_id>": {"strengths": ["..."], "weaknesses": ["..."]}}, "research_gap": ["..."] }',
    '',
    'Requirements:',
    '- method_evolution: 2-5 sentences tracing how the technical route evolved across these papers.',
    '- strengths_weaknesses: per paper_id, 2-4 strengths and 2-4 weaknesses.',
    '- research_gap: 3-6 concrete open problems synthesized from the papers.',
    '',
    '=== PAPERS ===',
  ];
  for (const s of summaries) {
    lines.push(`\n[${s.paper_id}] ${s.title} (${s.year})`);
    lines.push(`  research_problem: ${truncate(s.research_problem, 500)}`);
    lines.push(`  proposed_method: ${truncate(s.proposed_method, 500)}`);
    lines.push(`  limitations: ${(s.limitations || []).join('; ').slice(0, 400)}`);
    lines.push(`  research_gap: ${(s.research_gap || []).join('; ').slice(0, 400)}`);
  }
  return lines.join('\n');
}

async function comparePapers(paperIds, opts) {
  opts = opts || {};
  const summaries = paperIds.map(paperSummary);
  const provider = getProvider(opts.provider, opts.model);
  const result = { provider: provider.name, papers: summaries, method_evolution: '', strengths_weaknesses: {}, research_gap: [] };

  if (provider.name !== 'heuristic' && provider.available()) {
    try {
      const resp = await provider.complete(
        'You are a senior research analyst comparing academic papers.',
        buildComparisonPrompt(summaries)
      );
      const parsed = extractJson(resp.text);
      if (parsed && typeof parsed === 'object') {
        result.method_evolution = parsed.method_evolution || '';
        result.strengths_weaknesses = parsed.strengths_weaknesses || {};
        result.research_gap = parsed.research_gap || [];
        result.provider = provider.name;
        result.model = resp.model || opts.model || '';
        return result;
      }
    } catch (e) { /* fall through */ }
  }

  result.provider = 'heuristic';
  result.method_evolution = summaries.map(s => `${s.title}（${s.year}）提出 ${truncate(s.proposed_method, 120) || 'N/A'}`).join(' → ');
  result.strengths_weaknesses = {};
  for (const s of summaries) {
    result.strengths_weaknesses[s.paper_id] = {
      strengths: s.contributions.length ? s.contributions : ['未识别'],
      weaknesses: s.limitations.length ? s.limitations : ['未识别'],
    };
  }
  const gaps = [];
  for (const s of summaries) { gaps.push(...(s.research_gap || []), ...(s.limitations || [])); }
  result.research_gap = [...new Set(gaps.filter(Boolean))].slice(0, 8);
  return result;
}

module.exports = { comparePapers };

File: backend-fastapi/app/services/comparer.py
"""多篇论文对比：技术路线演化、优缺点、Research Gap。"""
from __future__ import annotations

from typing import Any, Dict, List, Optional

from ..llm import get_provider
from ..llm.base import ProviderError
from ..store import store
from .common import extract_json, truncate


def _paper_summary(paper_id: str) -> Dict[str, Any]:
    meta = store.get_metadata(paper_id)
    analysis = store.get_analysis(paper_id) or {}
    sec = analysis.get("sections", {}) or {}
    return {
        "paper_id": paper_id,
        "title": meta.get("title") or analysis.get("title") or paper_id,
        "year": meta.get("year") or analysis.get("year"),
        "research_problem": sec.get("research_problem", ""),
        "proposed_method": sec.get("proposed_method", ""),
        "results": sec.get("results", ""),
        "contributions": sec.get("contributions", []),
        "limitations": sec.get("limitations", []),
        "research_gap": sec.get("research_gap", []),
    }


def _build_comparison_prompt(summaries: List[Dict[str, Any]]) -> str:
    lines = ["Compare the following papers and return ONLY valid JSON (no fences) matching this schema:",
             '{ "method_evolution": "...", '
             '"strengths_weaknesses": {"<paper_id>": {"strengths": ["..."], "weaknesses": ["..."]}}, '
             '"research_gap": ["..."] }',
             "",
             "Requirements:",
             "- method_evolution: 2-5 sentences tracing how the technical route evolved across these papers.",
             "- strengths_weaknesses: per paper_id, 2-4 strengths and 2-4 weaknesses.",
             "- research_gap: 3-6 concrete open problems synthesized from the papers.",
             "",
             "=== PAPERS ==="]
    for s in summaries:
        lines.append(f"\n[{s['paper_id']}] {s['title']} ({s['year']})")
        lines.append(f"  research_problem: {truncate(s['research_problem'], 500)}")
        lines.append(f"  proposed_method: {truncate(s['proposed_method'], 500)}")
        lines.append(f"  limitations: {'; '.join(s['limitations'])[:400]}")
        lines.append(f"  research_gap: {'; '.join(s['research_gap'])[:400]}")
    return "\n".join(lines)


def compare_papers(paper_ids: List[str], provider_name: Optional[str] = None, model: Optional[str] = None) -> Dict[str, Any]:
    summaries = [_paper_summary(pid) for pid in paper_ids]
    provider = get_provider(provider_name, model)

    result: Dict[str, Any] = {
        "provider": provider.name,
        "papers": summaries,
        "method_evolution": "",
        "strengths_weaknesses": {},
        "research_gap": [],
    }

    if provider.name != "heuristic":
        try:
            resp = provider.complete(
                system="You are a senior research analyst comparing academic papers.",
                user=_build_comparison_prompt(summaries),
            )
            parsed = extract_json(resp.text)
            if isinstance(parsed, dict):
                result["method_evolution"] = parsed.get("method_evolution", "")
                result["strengths_weaknesses"] = parsed.get("strengths_weaknesses", {})
                result["research_gap"] = parsed.get("research_gap", [])
                result["provider"] = provider.name
                result["model"] = resp.model or model or ""
                return result
        except ProviderError:
            pass

    # heuristic 回退
    result["provider"] = "heuristic"
    result["method_evolution"] = " → ".join(
        f"{s['title']}（{s['year']}）提出 {truncate(s['proposed_method'], 120) or 'N/A'}" for s in summaries
    )
    result["strengths_weaknesses"] = {
        s["paper_id"]: {
            "strengths": s["contributions"] or ["未识别"],
            "weaknesses": s["limitations"] or ["未识别"],
        }
        for s in summaries
    }
    gaps: List[str] = []
    for s in summaries:
        gaps.extend(s["research_gap"])
        gaps.extend(s["limitations"])
    seen: set = set()
    result["research_gap"] = [g for g in gaps if g and not (g in seen or seen.add(g))][:8]
    return result
